File: solutions/javascript/protein-translation/1/protein-translation.js
//
// This is only a SKELETON file for the 'Protein Translation' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

 const peptideMap = {
  "AUG": "Methionine",
  "UUU": "Phenylalanine",
  "UUC": "Phenylalanine",
  "UUA": "Leucine",
  "UUG": "Leucine",
  "UCU": "Serine",
  "UCC": "Serine",
  "UCA": "Serine",
  "UCG": "Serine",
  "UAU": "Tyrosine",
  "UAC": "Tyrosine",
  "UGU": "Cysteine",
  "UGC": "Cysteine",
  "UGG": "Tryptophan",
  "UAA": "STOP",
  "UAG": "STOP",
  "UGA": "STOP"
 };
  

export const translate = (sequence) => {
  let res =  []
  if(sequence==undefined)
    return res
  for(let i=0; i<sequence.length; i+=3){
    let nuc = sequence.substring(i,i+3)
    let amino = peptideMap[nuc]
    if(amino==undefined)
      throw new Error("Invalid codon");
    if(amino=="STOP")
      break;
    res.push(amino)
  }
  return res
};

File: solutions/dart/protein-translation/1/lib/protein_translation.dart
import 'dart:math';

class ProteinTranslation {
  static Map<String,String> codon2amino = {
    'AUG' : 'Methionine',
    'UUU' : 'Phenylalanine',
    'UUC' : 'Phenylalanine',
    'UUA' : 'Leucine',
    'UUG' : 'Leucine',
    'UCU' : 'Serine',
    'UCC' : 'Serine',
    'UCA' : 'Serine',
    'UCG' : 'Serine',
    'UAU' : 'Tyrosine',
    'UAC' : 'Tyrosine',
    'UGU' : 'Cysteine',
    'UGC' : 'Cysteine',
    'UGG' : 'Tryptophan',
    'UAA' : 'STOP',
    'UAG' : 'STOP',
    'UGA' : 'STOP'
  };
  
  List<String> translate(String sequence){
    List<String> polypeptide = [];
    int i=0;
    bool stopped = false;
    for(var i=0; i<=sequence.length-3; i+=3){
      String codon = sequence.substring(i, i+3);
      if(codon2amino.containsKey(codon)){
        String translation = codon2amino[codon]!;
        if(translation == 'STOP'){
          stopped = true;
          break;
        }
        else
          polypeptide.add(translation);
      }
      else 
        throw ArgumentError('invalid codon');  
    }
    if(!stopped && sequence.length%3!=0)
      throw ArgumentError('incomplete sequence');  
    return polypeptide;
  }
}
